bpfassets/perf_event/perf_event.c

```diff
@@ -188,7 +188,9 @@ static inline u64 get_on_cpu_avg_freq(u32 *cpu_id, u64 on_cpu_cycles_delta, u64
 // int kprobe__finish_task_switch(switch_args *ctx)
 int kprobe__finish_task_switch(struct pt_regs *ctx, struct task_struct *prev)
 {
-    u64 cur_pid = bpf_get_current_pid_tgid() >> 32;
+    // u64 cur_pid = bpf_get_current_pid_tgid() & 0xffffff;
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    u32 cur_pid = task->tgid;
 #ifdef SET_GROUP_ID
     u64 cgroup_id = bpf_get_current_cgroup_id();
```

Comment:
We probably cannot use the current cgroup id, but we should also use the previous cgroup id!
This would require a ebpf hash to track the previous cgroup id in the given cpu.

Reply:
This is also the case of the command (wrongly reported). However, since we aggregate all system processes, we probably do not need the command. So we could remove it to minimize overheads.

Reply:
Based on the info below, the function runs in the new thread context, so `cgroup_id = bpf_get_current_cgroup_id();` should get the new pid's cgroup.

```The finish_task_switch() function is called in the context of the next-running thread. An eBPF program can instrument this function and argument using kprobes, fetch the current PID (via bpf_get_current_pid_tgid()), and also fetch a high resolution timestamp (bpf_ktime_get_ns())```

Reply:
Right, since "`the finish_task_switch() function is called in the context of the next-running thread` ", and the collected metrics are for the previous thread, we need to use the pid and cgroup id of the previous thread.